### ISSUE TITLE
Broken pipe error when running object_storage_service benchmark

### ISSUE THREAD

**Issue by hungvelo** [GoogleCloudPlatform/PerfKitBenchmarker#136]
Hi,
Has anyone seen this error, Broken pipe, during benchmark object_storage_service?
./pkb.py --project=elite-firefly-786 --benchmarks=object_storage_service --machine_type=n1-standard-2
pkb.log
...
2015-02-25 21:38:46,569 vm_util.py:335 INFO     Running: /usr/bin/ssh -A -p 22 perfkit@104.154.77.220 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /tmp/perfkitbenchmarker/run_fcd580f8/perfkitbenchmarker_keyfile /scratch0/run/object_storage_api_tests.py --bucket=pkbfcd580f8 --storage_provider=GCS --iterations=200 --scenario=ListConsistency
2015-02-26 01:34:05,081 vm_util.py:349 DEBUG    Ran /usr/bin/ssh -A -p 22 perfkit@104.154.77.220 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /tmp/perfkitbenchmarker/run_fcd580f8/perfkitbenchmarker_keyfile /scratch0/run/object_storage_api_tests.py --bucket=pkbfcd580f8 --storage_provider=GCS --iterations=200 --scenario=ListConsistency. Got return code (0). STDOUT: STDERR: Warning: Permanently added '104.154.77.220' (ECDSA) to the list of known hosts.^M
…..
Final stats:
INFO:root:list-after-write consistency percentage: 87.000000
INFO:root:list-after-write inconsistency window: {"average": 39.82011866569519, "p1": 1.0820820331573486, "p5": 1.132835865020752, "p50": 5.805269002914429, "p90": 134.60191202163696, "p99": 283.1713089942932, "p99.9": 283.1713089942932, "stddev": 73.03022897677377}
INFO:root:list-after-write latency: {"average": 1.2659037058380829, "p1": 0.9758129119873047, "p5": 1.0399689674377441, "p50": 1.1664412021636963, "p90": 1.4518799781799316, "p99": 5.052584886550903, "p99.9": 5.406584978103638, "stddev": 0.4693329095847918}
INFO:root:list-after-update consistency percentage: 89.000000
INFO:root:list-after-update inconsistency window: {"average": 21.084003946997903, "p1": 1.2440087795257568, "p5": 1.4276509284973145, "p50": 8.543776035308838, "p90": 53.372872829437256, "p99": 147.42951703071594, "p99.9": 147.42951703071594, "stddev": 34.4644712034284}
INFO:root:list-after-update latency: {"average": 1.263827731100361, "p1": 0.9965829849243164, "p5": 1.017981767654419, "p50": 1.1851608753204346, "p90": 1.4783661365509033, "p99": 5.225747108459473, "p99.9": 5.397141933441162, "stddev": 0.4733233401719766}

2015-02-26 01:34:05,095 pkb.py:256 ERROR    Error during benchmark object_storage_service
Traceback (most recent call last):
  File "/root/PerfKitBenchmarker/perfkitbenchmarker/pkb.py", line 238, in RunBenchmark
    samples = benchmark.Run(benchmark_specification)
  File "/root/PerfKitBenchmarker/perfkitbenchmarker/benchmarks/object_storage_service_benchmark.py", line 806, in Run
    print results
IOError: [Errno 32] Broken pipe
2015-02-26 01:34:05,096 vm_util.py:335 INFO     Running: gcloud compute instances delete perfkit-fcd580f8-0 --project elite-firefly-786 --format json --quiet --zone us-central1-a
2015-02-26 01:34:37,261 vm_util.py:349 DEBUG    Ran gcloud compute instances delete perfkit-fcd580f8-0 --project elite-firefly-786 --format json --quiet --zone us-central1-a. Got return code (0). STDOUT: []
STDERR: Deleted [https://www.googleapis.com/compute/v1/projects/elite-firefly-786/zones/us-central1-a/instances/perfkit-fcd580f8-0].

**Comment by voellm** [comment]
+Stanley

On Fri, Feb 27, 2015 at 12:29 AM, Hung <notifications@github.com> wrote:

> Hi,
> Has anyone seen this error, Broken pipe, during benchmark
> object_storage_service?
> ./pkb.py --project=elite-firefly-786 --benchmarks=object_storage_service
> --machine_type=n1-standard-2
> pkb.log
> ...
> 2015-02-25 21:38:46,569 vm_util.py:335 INFO Running: /usr/bin/ssh -A -p 22
> perfkit@104.154.77.220 -2 -o UserKnownHostsFile=/dev/null -o
> StrictHostKeyChecking=no -o IdentitiesOnly=yes -o
> PreferredAuthentications=publickey -o PasswordAuthentication=no -o
> ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o
> ServerAliveCountMax=10 -i
> /tmp/perfkitbenchmarker/run_fcd580f8/perfkitbenchmarker_keyfile
> /scratch0/run/object_storage_api_tests.py --bucket=pkbfcd580f8
> --storage_provider=GCS --iterations=200 --scenario=ListConsistency
> 2015-02-26 01:34:05,081 vm_util.py:349 DEBUG Ran /usr/bin/ssh -A -p 22
> perfkit@104.154.77.220 -2 -o UserKnownHostsFile=/dev/null -o
> StrictHostKeyChecking=no -o IdentitiesOnly=yes -o
> PreferredAuthentications=publickey -o PasswordAuthentication=no -o
> ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o
> ServerAliveCountMax=10 -i
> /tmp/perfkitbenchmarker/run_fcd580f8/perfkitbenchmarker_keyfile
> /scratch0/run/object_storage_api_tests.py --bucket=pkbfcd580f8
> --storage_provider=GCS --iterations=200 --scenario=ListConsistency. Got
> return code (0). STDOUT: STDERR: Warning: Permanently added
> '104.154.77.220' (ECDSA) to the list of known hosts.^M
> .....
> Final stats:
> INFO:root:list-after-write consistency percentage: 87.000000
> INFO:root:list-after-write inconsistency window: {"average":
> 39.82011866569519, "p1": 1.0820820331573486, "p5": 1.132835865020752,
> "p50": 5.805269002914429, "p90": 134.60191202163696, "p99":
> 283.1713089942932, "p99.9": 283.1713089942932, "stddev": 73.03022897677377}
> INFO:root:list-after-write latency: {"average": 1.2659037058380829, "p1":
> 0.9758129119873047, "p5": 1.0399689674377441, "p50": 1.1664412021636963,
> "p90": 1.4518799781799316, "p99": 5.052584886550903, "p99.9":
> 5.406584978103638, "stddev": 0.4693329095847918}
> INFO:root:list-after-update consistency percentage: 89.000000
> INFO:root:list-after-update inconsistency window: {"average":
> 21.084003946997903, "p1": 1.2440087795257568, "p5": 1.4276509284973145,
> "p50": 8.543776035308838, "p90": 53.372872829437256, "p99":
> 147.42951703071594, "p99.9": 147.42951703071594, "stddev": 34.4644712034284}
> INFO:root:list-after-update latency: {"average": 1.263827731100361, "p1":
> 0.9965829849243164, "p5": 1.017981767654419, "p50": 1.1851608753204346,
> "p90": 1.4783661365509033, "p99": 5.225747108459473, "p99.9":
> 5.397141933441162, "stddev": 0.4733233401719766}
>
> 2015-02-26 01:34:05,095 pkb.py:256 ERROR Error during benchmark
> object_storage_service
> Traceback (most recent call last):
> File "/root/PerfKitBenchmarker/perfkitbenchmarker/pkb.py", line 238, in
> RunBenchmark
> samples = benchmark.Run(benchmark_specification)
> File
> "/root/PerfKitBenchmarker/perfkitbenchmarker/benchmarks/object_storage_service_benchmark.py",
> line 806, in Run
> print results
> IOError: [Errno 32] Broken pipe
> 2015-02-26 01:34:05,096 vm_util.py:335 INFO Running: gcloud compute
> instances delete perfkit-fcd580f8-0 --project elite-firefly-786 --format
> json --quiet --zone us-central1-a
> 2015-02-26 01:34:37,261 vm_util.py:349 DEBUG Ran gcloud compute instances
> delete perfkit-fcd580f8-0 --project elite-firefly-786 --format json --quiet
> --zone us-central1-a. Got return code (0). STDOUT: []
> STDERR: Deleted [
> https://www.googleapis.com/compute/v1/projects/elite-firefly-786/zones/us-central1-a/instances/perfkit-fcd580f8-0
> ].
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/136>.
>



-- 
Anthony F. Voellm (aka Tony)
Google Voice:  (650) 516-7382
<https://www.google.com/voice/b/0?pli=1#phones>
Blog: http://perfguy.blogspot.com/

**Comment by stfeng** [comment]
From the log provided, the test has completed successfully, but when it tries to print the result to std out, it got an broken pipe error, which typically happens if your standard out is to a terminal window which you have just logged-out or timedout.

Make sure you don't log out of your shell window that you started this test, or redirect your standard out to a file.

**Comment by hungvelo** [comment]
But I ran it with nohup and redirected the output std error and std output to a file as follow:
nohup ./pkb.py --project=elite-firefly-786 --benchmarks=object_storage_service --machine_type=n1-standard-2 2>&1 | tee object_storage_service.out &
This works fine with other benchmarks where the PerfKitBenchmark Results Summary was printed out successfully to the output file, even when the ssh connection was terminated.

**Comment by stfeng** [comment]
This issue is pretty simple: we failed to write to stdout when doing a simple "print result" from the test. Make sure the stdout is robust and accessible for the duration of this test will fix this problem. Perhaps this is just a transient error, but I am not sure. 

I have been using redirects all the time on this benchmark as well and have been running it for weeks, never seen this problem.

Just for trying, let's not use tee command here, but just write to the file. (you can always do tail -f on the file to see the real time results)

**Comment by hungvelo** [comment]
Thanks, Stanley. With your first explanation, I went ahead and disabled SSH timeout (I have root access), and this time the test took much longer (>10 hrs) and completed successfully with the perfkitbenchmarker_results.json file generated. There's still one error at the very end when it's trying to clean up and remove the bucket:
Traceback (most recent call last):
  File "./pkb.py", line 21, in <module>
    sys.exit(Main())
  File "/root/PerfKitBenchmarker/perfkitbenchmarker/pkb.py", line 358, in Main
    return RunBenchmarks()
  File "/root/PerfKitBenchmarker/perfkitbenchmarker/pkb.py", line 324, in RunBenchmarks
    RunBenchmark(benchmark, collector)
  File "/root/PerfKitBenchmarker/perfkitbenchmarker/pkb.py", line 244, in RunBenchmark
    benchmark.Cleanup(benchmark_specification)
  File "/root/PerfKitBenchmarker/perfkitbenchmarker/benchmarks/object_storage_service_benchmark.py", line 819, in Cleanup
    OBJECT_STORAGE_BENCHMARK_DICTIONARY[FLAGS.storage].Cleanup(vms[0])
  File "/root/PerfKitBenchmarker/perfkitbenchmarker/benchmarks/object_storage_service_benchmark.py", line 722, in Cleanup
    DeleteBucketWithRetry(vm, remove_content_cmd, remove_bucket_cmd)
  File "/root/PerfKitBenchmarker/perfkitbenchmarker/benchmarks/object_storage_service_benchmark.py", line 351, in DeleteBucketWithRetry
    raise BucketRemovalError('Failed to remove the bucket')
perfkitbenchmarker.benchmarks.object_storage_service_benchmark.BucketRemovalError: Failed to remove the bucket

**Comment by hungvelo** [comment]
Reviewing the log, I saw a lot of the same error (number of attempts 120) below:
2015-02-27 03:28:10,893 ERROR    Failed to perform the removal op. Number of attempts: 1. Error is Got non-zero return code (1) executing /home/perfkit/google-cloud-sdk/bin/gsutil -m rm -r gs://pkbb260fddb/*
Full command: /usr/bin/ssh -A -p 22 perfkit@104.197.1.20 -2 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -o PreferredAuthentications=publickey -o PasswordAuthentication=no -o ConnectTimeout=5 -o GSSAPIAuthentication=no -o ServerAliveInterval=30 -o ServerAliveCountMax=10 -i /tmp/perfkitbenchmarker/run_b260fddb/perfkitbenchmarker_keyfile /home/perfkit/google-cloud-sdk/bin/gsutil -m rm -r gs://pkbb260fddb/*
STDOUT: STDERR: Warning: Permanently added '104.197.1.20' (ECDSA) to the list of known hosts.^M
Removing gs://pkbb260fddb/file-11.dat#1425002760376000...
...
2015-02-27 04:46:21,199 ERROR    Failed to perform the removal op. Number of attempts: 117. Error is Got non-zero return code (1) executing /home/perfkit/google-cloud-sdk/bin/gsutil rb gs://pkbb260fddb
2015-02-27 04:46:52,775 ERROR    Failed to perform the removal op. Number of attempts: 118. Error is Got non-zero return code (1) executing /home/perfkit/google-cloud-sdk/bin/gsutil rb gs://pkbb260fddb
2015-02-27 04:47:24,483 ERROR    Failed to perform the removal op. Number of attempts: 119. Error is Got non-zero return code (1) executing /home/perfkit/google-cloud-sdk/bin/gsutil rb gs://pkbb260fddb
2015-02-27 04:47:56,362 ERROR    Failed to perform the removal op. Number of attempts: 120. Error is Got non-zero return code (1) executing /home/perfkit/google-cloud-sdk/bin/gsutil rb gs://pkbb260fddb
2015-02-27 04:48:26,393 ERROR    Exceeded max retry limit for removing the empty bucket
2015-02-27 04:48:26,393 ERROR    Error during benchmark object_storage_service
    raise BucketRemovalError('Failed to remove the bucket')
BucketRemovalError: Failed to remove the bucket
    raise BucketRemovalError('Failed to remove the bucket')
perfkitbenchmarker.benchmarks.object_storage_service_benchmark.BucketRemovalError: Failed to remove the bucket

**Comment by stfeng** [comment]
Right, this is a known issue with the cloud provider "GCP" where a 503 service side error would be returned when we attempt to delete an EMPTY bucket. We added retries up to 120 times to tolerate this but apparently and unfortunately in this particular case, after many hours and 120 times of retrying, the bucket is still not removed.

The good news is there is no limit of number of buckets you can have on GCS, so this won't block you from creating new buckets later. 

The mitigation is wait for a few more days and come back trying to delete this bucket manually using gsutil rb "bucket name"

**Comment by hungvelo** [comment]
Doesn't this effect "End to End Runtime" with 120 times of retries? Actually, it did not even print out End to End Runtime.

**Comment by stfeng** [comment]
It's the balance between proper resource cleanup and test runtime. We initially had this test without much wait and simply bail when we failed to cleanup the bucket, but that left lots of resource hanging around which isn't very good either (e.g., some provider enforces a limit of max 30 buckets). With added wait and tenacious retry, we were able to successfully clean-up for almost all cases but there is still a very long tail of cases that requires even longer wait, for those, we have to err on the side of having a "reasonable" test runtime and bail.
